### PR TITLE
Fix weaviate object type metadata

### LINKED ISSUE #908: Weaviate properties does not allow for object data type
[Current implementation of batch object insert](https://github.com/hwchase17/langchainjs/blob/b45313c85437df139bb67f35313de69f77beb2a8/langchain/src/vectorstores/weaviate.ts#L62) to Weaviate allows all data type to be inserted. This will result in error when inserting `Documents` that have `Object` type in their `metadata`. For example the `metadata` of `Document` returned by `PDFLoader` looks like,

PDFLoader metadata example

P.S. `PDFLoader` metadata output are `$.pdf`, `$.loc`, and `$.source`

And when those `Document` with `Object` metadata are inserted to Weaviate using the current implementation, none of the `Document` will be stored in Weaviate. Further investigation, Weaviate do output error when inserting the `Document`, as you can see down here,

Weaviate error output

Seeing this error, I think the current implementation are lacking,

1. Metadata filtering of unsupported data ([currently supported data type by Weaviate](https://weaviate.io/developers/weaviate/config-refs/datatypes#introduction) ), so that any non supported data type will be filtered or raised error (maybe add config to define the behaviour of unsupported data?)
2. Raised error sent by Weaviate on addVectors (maybe return the error sent, because on batch there could be some successful document insertion as well)

### LINKED ISSUE #926: Weaviate Vectore Store: Error "no graphql provider present, this is most likely because no schema is present. Import a schema first!"
I am trying to get started with Weaviate via LangChainJS, for the first time, and I am migrating my code from using Pinecone.

In my code below I am creating some documents from a PDF and using embeddings from open AI to import the documents to Weaviate. I am just running Weaviate in docker locally with a basic default docker-compose from their generator in their docs site.

I get no errors at all from `WeaviateStore.fromDocuments` but I can not retrieve any data as it seems no schema has been generated after importing the data.

I get the error from `store.similaritySearch`:
`no graphql provider present, this is most likely because no schema is present. Import a schema first`
```bash
'Error in similaritySearch' Error: GraphQL Error (Code: 422): 
{"response":
{"error":[{"message":"no graphql provider present, this is most likely because no schema is present. 
Import a schema first!"}],"status":422,"headers":{}},
"request":{"query":"{Get{pdf-test(nearVector:{vector:
[-0.025541332,-0.013953761,0.01518418,-0.012939679,-0.014792069,0.007930117,-0.023
...
```

It is my assumption that schemas are auto generated based on the data I put into Weaviate, sic :  https://weaviate.io/developers/weaviate/configuration/schema-configuration#auto-schema

Here is my code, if anyone can help me figure out what I am doing wrong:

```typescript
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { OpenAIEmbeddings } from 'langchain/embeddings/openai';
import { WeaviateStore } from 'langchain/vectorstores/weaviate';
import { client } from '@/utils/weaviate-client';
import { CustomPDFLoader } from '@/utils/customPDFLoader';
import { WEAVIATE_INDEX_NAME } from '@/config/weaviate';
import { DirectoryLoader } from 'langchain/document_loaders/fs/directory';

/* Name of directory to retrieve your files from */
const filePath = 'public/docs';

export const run = async () => {
  try {
    /*load raw docs from the all files in the directory */
    const directoryLoader = new DirectoryLoader(filePath, {
      '.pdf': (path) => new CustomPDFLoader(path),
    });

    // const loader = new PDFLoader(filePath);
    const rawDocs = await directoryLoader.load();

    /* Split text into chunks */
    const textSplitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 200,
    });

    const docs = await textSplitter.splitDocuments(rawDocs);
    console.log('split docs', docs);

    console.log('creating vector store...');
    /*create and store the embeddings in the vectorStore*/
    const embeddings = new OpenAIEmbeddings();

    //embed the PDF documents
    const store = await WeaviateStore.fromDocuments(docs, embeddings, {
      client,
      indexName: WEAVIATE_INDEX_NAME,
      textKey: 'text',
      metadataKeys: ['source', 'pdf_numpages', 'loc'],
    });

    const results = await store.similaritySearch('broadband', 1);
    console.log(results);
  } catch (error) {
    console.log('error', error);
    throw new Error('Failed to ingest your data');
  }
};

(async () => {
  await run();
  console.log('ingestion complete');
})();

```

Here is my docker compose if it helps:

```yaml
---
version: '3.4'
services:
  weaviate:
    command:
    - --host
    - 0.0.0.0
    - --port
    - '8080'
    - --scheme
    - http
    image: semitechnologies/weaviate:1.18.3
    ports:
    - 8080:8080
    restart: on-failure:0
    environment:
      OPENAI_APIKEY: **REDACTED**
      QUERY_DEFAULTS_LIMIT: 25
      AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED: 'true'
      PERSISTENCE_DATA_PATH: '/var/lib/weaviate'
      DEFAULT_VECTORIZER_MODULE: 'text2vec-openai'
      ENABLE_MODULES: 'text2vec-openai'
      CLUSTER_HOSTNAME: 'node1'
      AUTOSCHEMA_ENABLED: 'true'
    volumes:
      - ./weaviate:/var/lib/weaviate
...

```

Here are the relevant env vars:

```bash
WEAVIATE_INDEX_NAME=pdf-test
WEAVIATE_SCHEME=http
WEAVIATE_HOST=127.0.0.1:8080

```

### PR DESCRIPTION
**Pull request by fauh45**
This pull request will add the following to the weaviate vectorstore implementation in langchain,

- Flatten document metadata before inserting to Weaviate
- Add integration test to test out deep document metadata

It should fix and closes #908 also closes #926

### CODE REVIEW

File: langchain/src/util/flatten.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const flattenObject = (obj: Record<string, any>, deep = true) => {
```
[the remark this answers is not in the document]
Comment (nfcampos) [reply langchain/src/util/flatten.ts:2]: Thanks for reporting, this is fixed here https://github.com/hwchase17/langchainjs/pull/1086

[the remark this answers is not in the document]
Comment (harshtomar6) [reply langchain/src/util/flatten.ts:2]: @nfcampos Thanks for the quick fix. However, I also found that weaviate allows this regex for their metadata keys `/[_A-Za-z][_0-9A-Za-z]*/` 
So flattening object with `key1.key2` won't work. You have to do something like `key1_key2`. I verified this locally. 
Documents were not actually loaded before and were loaded only after I changed it the latter

[the remark this answers is not in the document]
Comment (nfcampos) [reply langchain/src/util/flatten.ts:2]: Ah thanks, fixed that now

[the remark this answers is not in the document]
Comment (fauh45) [reply langchain/src/util/flatten.ts:2]: Oof, sorry for not thinking that through. Thanks @nfcampos for fixing it

**Pull request merged**